skbio/stats/composition.py

```diff
@@ -608,10 +612,274 @@ def centralize(mat):
+          a novel method for studying microbial composition",
+          Microbial Ecology in Health & Disease, (2015), 26
+
+    Examples
+    --------
+    First import all of the necessary modules
+
+    >>> from skbio.stats.composition import ancom
+    >>> import pandas as pd
+
+    Now lets load in a pandas dataframe with sample and feature ids
+    for our data matrix.
+    >>> table = pd.DataFrame(
+    ...     [[10., 11., 10., 10., 10., 10., 10.],
+    ...      [10.5, 11.5, 10.5, 10.5, 10.5, 10.5, 10.5],
+    ...      [10., 11., 10., 10., 10., 10., 10.],
+    ...      [20., 21., 10., 10., 10., 10., 10.],
+    ...      [20.5, 21.5, 10.5, 10.5, 10.5, 10.5, 10.5],
+    ...      [20.3, 21.3, 10.2, 10.3, 10.1, 10.6, 10.4]],
+    ...     index=['s1','s2','s3','s4','s5','s6'],
+    ...     columns=['b1','b2','b3','b4','b5','b6','b7'])
+
+    Then create a create a category vector.  In this scenerio, there
+    are only two classes, so the first three samples fall under the first
+    class while the last three samples fall under the last class
+    >>> cats = pd.Series([0, 0, 0, 1, 1, 1],
+    ...                  index=['s1','s2','s3','s4','s5','s6'])
+
+    Now run ancom and see if there are any features that have any
+    significant differences
+    >>> results = ancom(table, cats)
+    >>> results['W']
+    b1    6
+    b2    6
+    b3    2
+    b4    2
+    b5    2
+    b6    2
+    b7    2
+    Name: W, dtype: int64
+    >>> results['reject']
+    b1     True
+    b2     True
+    b3    False
+    b4    False
+    b5    False
+    b6    False
+    b7    False
+    Name: reject, dtype: bool
+    """
+    if len(matrix) != len(categories):
+        raise ValueError('The number of samples in mat needs' +
+                         'to be the same as the number of samples' +
+                         'in cats')
+    mat = _check_composition(matrix, ignore_zeros=False)
+    cats = pd.Series(categories)
+
+    mat.sort_index(inplace=True)
+    cats = cats.sort_index()
+    labs = mat.columns
+
+    n_samp, n_feat = mat.shape
+
+    _logratio_mat = _log_compare(mat.values, cats.values, significance_test)
+    logratio_mat = _logratio_mat + _logratio_mat.T
+
+    # Multiple comparisons
+    if multiple_comparisons_correction is not None:
+        if multiple_comparisons_correction == 'holm-bonferroni':
+            logratio_mat = np.apply_along_axis(_holm_bonferroni,
+                                               1, logratio_mat)
+
+    np.fill_diagonal(logratio_mat, 1)
+
+    W = (logratio_mat < alpha).sum(axis=1)
+    c_start = max(W) / n_feat
+    if c_start < theta:
+        reject = np.array([False] * len(W))
+        return pd.DataFrame({'W': pd.Series(W, index=labs),
+                             'reject': pd.Series(reject, index=labs)})
+
+    cutoff = c_start - np.linspace(0.05, 0.25, 5)
+    dels = np.zeros_like(cutoff)
+    prop_cut = np.zeros_like(cutoff, dtype=np.float32)
+    for cut in range(len(cutoff)):
```

Comment:
Oh, I see. Lemme think about it too...
On Oct 22, 2015 12:49 PM, "mortonjt" <notifications@github.com> wrote:

> In skbio/stats/composition.py
> <https://github.com/biocore/scikit-bio/pull/1114#discussion_r42789473>:
>
> > +            logratio_mat = np.apply_along_axis(_holm_bonferroni,
> > +                                               1, logratio_mat)
> > +
> > +    np.fill_diagonal(logratio_mat, 1)
> > +
> > +    W = (logratio_mat < alpha).sum(axis=1)
> > +    c_start = max(W) / n_feat
> > +    if c_start < theta:
> > +        reject = np.array([False] * len(W))
> > +        return pd.DataFrame({'W': pd.Series(W, index=labs),
> > +                             'reject': pd.Series(reject, index=labs)})
> > +
> > +    cutoff = c_start - np.linspace(0.05, 0.25, 5)
> > +    dels = np.zeros_like(cutoff)
> > +    prop_cut = np.zeros_like(cutoff, dtype=np.float32)
> > +    for cut in range(len(cutoff)):
>
> The problem is that W and cutoff are completely different dimensions, so
> to do this in one-line numpy broadcasting is required.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1114/files#r42789473>.
>